wicket-core/src/test/java/org/apache/wicket/markup/html/form/AutoLabelLabelProviderLocatorTest.java

```diff
@@ -0,0 +1,154 @@
+
+/**
+ * Tests {@code wicket:for} attribute functionality using {@link ILabelProviderLocator}
+ */
+@SuppressWarnings({ "rawtypes", "serial" })
+class AutoLabelLabelProviderLocatorTest extends WicketTestCase
+{
+	static class ILabelProviderLocatorPanel1 extends Panel implements ILabelProviderLocator
+	{
+
+		private TextField<String> text;
+
+		ILabelProviderLocatorPanel1(String id)
+		{
+			super(id);
+			WebMarkupContainer dummy = new WebMarkupContainer("dummy");
+			dummy.add(AttributeModifier.replace("class", "dummy1"));
+			add(dummy);
+			text = new TextField<>("text", Model.of(""));
+			dummy.add(text);
+		}
+
+		@Override
+		public Component getAutoLabelComponent()
+		{
+			return text;
+		}
+	}
+
+	static class ILabelProviderLocatorPanel2 extends Panel implements ILabelProviderLocator
+	{
+
+		private TextField<String> text;
+
+		ILabelProviderLocatorPanel2(String id)
+		{
+			super(id);
+			WebMarkupContainer dummy = new WebMarkupContainer("dummy");
+			dummy.add(AttributeModifier.replace("class", "dummy"));
+			add(dummy);
+
+			WebMarkupContainer dummy1 = new WebMarkupContainer("dummy1");
+			dummy1.add(AttributeModifier.replace("class", "dummy1"));
+			dummy.add(dummy1);
+
+			text = new TextField<>("text", Model.of(""));
+			dummy1.add(text);
+		}
+
+		@Override
+		public Component getAutoLabelComponent()
+		{
+			return text;
+		}
+	}
+
+	private interface IEditPanelProvider
+	{
+		IModel<String> getLabelText();
+
+		Panel createEditPanel(String id);
+	}
+
+	static class EditPage extends WebPage
+	{
+		EditPage(List<IEditPanelProvider> editPanelProviders)
+		{
+			Form<Void> form = new Form<>("form");
+			add(form);
+			RepeatingView editRow = new RepeatingView("edit-row");
+			form.add(editRow);
+			for (IEditPanelProvider panelProvider: editPanelProviders) {
+				WebMarkupContainer mc = new WebMarkupContainer(editRow.newChildId());
+				mc.add(new Label("label", panelProvider.getLabelText()));
+				mc.add(panelProvider.createEditPanel("edit-component"));
+				editRow.add(mc);
+			}
+		}
+	}
+
+	@Test
+	void markupIntoLabelInsertion()
+	{
+		List<IEditPanelProvider> providers = new ArrayList<>();
+
+		providers.add(new IEditPanelProvider()
+		{
+			@Override
+			public IModel<String> getLabelText()
+			{
+				return Model.of("Example1");
+			}
+
+			@Override
+			public Panel createEditPanel(String id)
+			{
+				return new ILabelProviderLocatorPanel1(id);
+			}
+		});
+
+		providers.add(new IEditPanelProvider()
+		{
+			@Override
+			public IModel<String> getLabelText()
+			{
+				return Model.of("Example2");
+			}
+
+			@Override
+			public Panel createEditPanel(String id)
+			{
+				return new ILabelProviderLocatorPanel2(id);
+			}
+		});
+
+		EditPage editPage = new EditPage(providers);
+		tester.startPage(editPage);
+		tester.assertRenderedPage(EditPage.class);
+		tester.assertContains("<span wicket:id=\"label\">Example1</span>");
+		tester.assertContains("<span wicket:id=\"label\">Example2</span>");
```

Comment:
shouldn't those assert that some specific `for="..."` attributes are generated instead of asserting their innerHTML ?

Reply:
Fixed